compiler/rustc_middle/src/ty/predicate/clause.rs

```diff
@@ -0,0 +1,28 @@
+use crate::ty::{
+    Const, ProjectionPredicate, RegionOutlivesPredicate, TraitPredicate, Ty, TypeOutlivesPredicate,
+};
+
+/// A clause is something that can appear in where bounds or be inferred
+/// by implied bounds.
+#[derive(Clone, Copy, PartialEq, Eq, Hash, TyEncodable, TyDecodable)]
+#[derive(HashStable, TypeFoldable, TypeVisitable, Lift)]
+pub enum Clause<'tcx> {
```

Comment:
lives with predicates

Reply:
It already is in `predicate` module.. Or do you think all of the predicate stuff should be in a single file? 🤔 

I feel like all of predicates are too much for a single file, it starts to get confusing/crowded...

Reply:
Yes, I think it should be in one file. It's not meaningful to treat all of these different predicate kinds as different modules, and I feell like it's not crowded if they're all really related.

Reply:
I'm iffy about that... Especially for types which have impls themselves (`ProjectionPredicate`, `InstantiatedPredicates`, ...), having a lot of impls for different types makes navigation noticeably harder imo.

Reply:
As someone who interacts with this code a lot, I strongly disagree.